### PR TITLE
VEText Migration - Update Twilio provider logic to use sms sender.

### PR DESCRIPTION
**Pull request by kalbfled**
# UPDATE:
Please see new the PR, this one is outdated and has unnecessary changes.
https://github.com/department-of-veterans-affairs/notification-api/pull/830

#747

POST requests to /v2/notifications/sms are used to send text messages to veterans.  In addition to the recipient's phone number or recipient_id, which can be used to look up the phone number, sending a request requires the sender's phone number.

These changes implement the ability to create text notifications using the value of *message_service_sid* in place of the sender's phone number when using Twilio as the SMS backend.  *messaging_service_sid*, if available, will be read from the *sms_sender_specifics* JSON field of a ServiceSmsSender instance associated with the authenticated service (user) and given template id, which is the other required field when making a request to this sms endpoint.

I found the existing code very difficult to follow.  Many of the edits in this PR are comments I made about types and purposes to facilitate not having to refer to different files frequently.  I also made formatting changes to appease the code scan test.

## Rough flow path for the endpoint
1. app/v2/notifications/post_notifications.py::post_notification
2. The above calls either *process_sms_or_email_notification* or *process_notification_with_recipient_identifier*.  The outcome is the creation of a new Notification instance from the JSON attributes contained in the POST request.
3. The notification is sent to *send_notification_to_queue* where it's added to the Celery Queue. 
4. The Celery worker processes each task by passing it to the "provider" (i.e. Twilio, etc.).  That is implemented in app/delivery/send_to_providers.py::send_sms_to_provider.

The `sms_sender_id` is also passed along the above flow so the specific one can be chosen when it's time to send the message.

### CODE REVIEW

File: app/clients/sms/twilio.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def send_sms(self, to, content, reference, sender=None):
         try:
             message = self._client.messages.create(
```
Comment (jakehova) [inline app/clients/sms/twilio.py:53]: This should be using message service id that gets passed in the request to Notify: https://www.twilio.com/docs/messaging/services 

```

 def send_sms(self, to, content, reference, sender=None):
        # could potentially select from potential numbers like this
        # from_number = random.choice(self._client.incoming_phone_numbers.list()).phone_number
        start_time = monotonic()        
        callback_url = "{}/notifications/sms/twilio/{}".format(
            self._callback_notify_url_host, reference) if self._callback_notify_url_host else ""
        try:
			if (sender is None):
				message = self._client.messages.create(
					to=to,
					from_=from_number,
					body=content,
					status_callback=callback_url,
				)
			else:
				 message = self._client.messages.create(
					to=to,
					body=content,
					messaging_service_sid=sender.message_service_id
					status_callback=callback_url
				)
            self.logger.info("Twilio send SMS request for {} succeeded: {}".format(reference, message.sid))
        except Exception as e:
            self.logger.error("Twilio send SMS request for {} failed".format(reference))
            raise e
        finally:
            elapsed_time = monotonic() - start_time
self.logger.info("Twilio send SMS request for {} finished in {}".format(reference, elapsed_time))

```

File: app/v2/notifications/post_notifications.py [hunk 2]
```diff
@@ -218,15 +229,15 @@ def process_sms_or_email_notification(*, form, notification_type, api_key, templ
         notification_type=notification_type,
         api_key_id=api_key.id,
         key_type=api_key.key_type,
-        client_reference=form.get('reference', None),
+        client_reference=form.get("reference"),
```
Comment (jakehova) [inline app/v2/notifications/post_notifications.py:232]: Why remove the default value of None?

**Review by jakehova — DISMISSED** [review]

File: app/v2/notifications/post_notifications.py (hunk 2 above)
Comment (kalbfled) [reply app/v2/notifications/post_notifications.py:232]: It's superfluous.  If you don't specify a default, None is the default.

File: app/clients/sms/twilio.py (hunk 1 above)
Comment (kalbfled) [reply app/clients/sms/twilio.py:53]: message_service_id, if available, is passed to send_sms as the "sender" parameter upstream in app/delivery/send_to_providers.py::send_sms_to_provider lines 82-95 (after these changes).

Comment (jakehova) [reply app/clients/sms/twilio.py:53]: right, this is the twilio.py file.  when you get that message service id, you need to create a new twilio message that specifies that message service id.  I saw the changes in send_sms_to_provider.  I wasnt sure how to properly access that property of sender, so I just made an example like above (sender.message_service_id in the else)

**Review by EvanParish — DISMISSED** [review]

**Comment by kalbfled** [comment]
This branch has been [successfully deployed](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/2870779107).

**Comment by k-macmillan** [comment]
I thoroughly appreciate you cleaning up the code. It's important to take a few minutes to clean things up when we find them and I love seeing it. This kind of thing saves us and future maintainers a lot of headache!

That said, I think `send_to_providers.py` needs a little more work. `send_sms_to_provider` was intended as a provider-agnostic method using polymorphism to call the relevant provider's `send_sms` method. What you have here breaks that, and allowing it to be commit to master is just asking for tech debt. I know how much work you put into this but I'm going to have to ask for a rework to include the provider-specific logic in the provider's class/method.

I would advise changing the existing `send_sms` methods to take `**kwargs` as their final parameter, then I'd include 
`notification.service_id` and `notification.reply_to_text` to the provider's `send_sms` method. Specifying `**kwargs` allows future changes to not require refactoring. Providers that have specifics (such as Twilio) would then call `dao_get_service_sms_sender_by_service_id_and_number` with the passed information. Calling it in the provider's `send_sms` method accomplishes the same goal you had before of only calling the `dao` method it if it was relevant. 

That last paragraph I leave at your discretion; it was a suggestion for how it could be handled. However you adjust it, please just ensure the whole method remains provider-agnostic.

**Comment by k-macmillan** [comment]
Briefly reviewed. This fails to meet the goal of the rework. `message_service_sid` is Twilio-specific and should not be in the `send_sms_to_provider` method. `send_sms_to_provider` should only pass additional information to the provider's `send_sms` method. If all goes well the only thing to change in `send_sms_to_provider` would be to add `notification.service_id` in the list of arguments being sent to `provider.send_sms()`. 

Inside the twilio client you can call `dao_get_service_sms_sender_by_service_id_and_number` and then set `messaging_service_sid`.

This is more a style thing, but if you wish to ditch the duplicated lines between 59 and 75 you could specify the values as a dictionary (`twilio_data` for instance), then pass `self._client.messages.create(**twilio_data)`, only adding the `messaging_service_sid` if necessary. 100% up to you.

**Review by kalbfled — COMMENTED** [review]
Master has been merged into this branch, thereby creating a diff with many items not related to the feature.  I recommend creating a new branch with a new PR.

**Comment by EvanParish** [comment]
Converting to draft. Will open new PR when code changes are in new branch since I merged master into this one.

**Pull request closed without merging**